src/ensembl/src/shared/hooks/useResizeObserver.ts

```diff
@@ -0,0 +1,68 @@
+// modified from https://github.com/ZeeCoder/use-resize-observer/blob/master/src/index.js
+
+import { useEffect, useState, useRef, useMemo, RefObject } from 'react';
+import windowService from 'src/services/window-service';
+
+type Params<T> = {
+  ref?: RefObject<T> | null;
+};
+
+export default function<T extends HTMLElement>(params: Params<T> = {}) {
+  const defaultRef = useRef(null);
+  const ref = params.ref || defaultRef;
+  const [size, setSize] = useState({ width: 0, height: 0 });
+
+  const sizeRef = useRef({
+    width: 0,
+    height: 0
+  });
+
+  useEffect(() => {
+    if (ref && !ref.current) {
+      return;
+    }
+
+    const element = ref.current;
+
+    if (!element) {
+      return;
+    }
+
+    const ResizeObserver = windowService.getResizeObserver();
+    console.log('ResizeObserver', ResizeObserver);
+    const resizeObserver = new ResizeObserver((entries) => {
+      if (!Array.isArray(entries)) {
+        return;
+      }
+
+      // We are only observing one element
+      const entry = entries[0];
+
+      if (!entry) {
+        return;
+      }
+
+      // `Math.round` is in line with how CSS resolves sub-pixel values
+      const newWidth = Math.round(entry.contentRect.width);
+      const newHeight = Math.round(entry.contentRect.height);
+      if (
+        sizeRef.current.width !== newWidth ||
+        sizeRef.current.height !== newHeight
+      ) {
+        sizeRef.current.width = newWidth;
+        sizeRef.current.height = newHeight;
+        setSize({ width: newWidth, height: newHeight });
+      }
+    });
+
+    resizeObserver.observe(element);
+
+    return () => resizeObserver.unobserve(element);
+  }, [ref]);
+
+  return useMemo(() => ({ ref, ...size }), [
```

Comment:
Do we need `useMemo` here?

Reply:
The author of the original library added `useMemo` in this commit: https://github.com/ZeeCoder/use-resize-observer/commit/5c4aa04e52f73d032f8136a59cec775ce40f7fb3#diff-1fdf421c05c1140f6d71444ea2b27638R69-R73

Hard to tell what problem he was trying to solve. The only difference is that now the hook will return the same object in certain situations rather than returning a new object all the time. Can't imagine why that should matter — building a new object is trivial, and React component shouldn't re-render If the return value from this hook is different from the previous return value. I suppose we could remove the useMemo.

Reply:
I was not able to be sure either. It it very unlikely that the dimensions will remain unchanged after playing around with the window size. Maybe we can remove it and look into adding it later on if it starts to cause any problem?

Reply:
Removed the useMemo